---
title: "Prepare the data"
author: "Meng Wang"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  \usepackage[utf8]{inputenc}
---

The purpose of this vignette demonstrates how to prepare data for ReSET. First, we read in and preprocess two 10X scRNAseq datasets containing around 3000 and 4000 Pan T cells from a single normal donor. The two datasets are available [here](https://support.10xgenomics.com/single-cell-gene-expression/datasets/2.1.0/t_3k) and [here](https://support.10xgenomics.com/single-cell-gene-expression/datasets/2.1.0/t_4k). We used `preprocess_data` function to filter and normalize the data. PCA were performed as well. The output is a Seurat object.

```{r message=FALSE, warning=FALSE, include=FALSE}
T3K_path <- system.file("extdata", '3k_pan_T_filtered_gene_bc_matrices/GRCh38', package = 'ReSET', mustWork = T)
T4K_path <- system.file("extdata", "4k_pan_T_filtered_gene_bc_matrices/GRCh38", package = 'ReSET', mustWork = T)

T3K_data <- ReSET::preprocess_data(T3K_path)
T4K_data <- ReSET::preprocess_data(T4K_path)
```

We performed graph-based clustering using `Seurat` and construct a tree of the clusters using hierarchical clustering on the mean of gene expression. Here is the dendrogram for `T3K` dataset.

```{r message=FALSE, warning=FALSE, fig.width = 6, fig.height = 3}
T3K_data <- Seurat::FindClusters(T3K_data, reduction.type = "pca", dims.use = 1:12, resolution = 2, print.output = 0) # 12 clusters
T3K_hierarchy <- ReSET::construct_hierarchy(t(T3K_data@scale.data[T3K_data@var.genes, ]), membership = T3K_data@ident)
stats:::plot.dendrogram(T3K_hierarchy$dend)
```

And here is the dendrogram for `T4K` dataset.

```{r message=FALSE, warning=FALSE, fig.width = 6, fig.height = 3}
T4K_data <- Seurat::FindClusters(T4K_data, reduction.type = "pca", dims.use = 1:12, resolution = 2, print.output = 0) # 13 clusters
T4K_hierarchy <- ReSET::construct_hierarchy(t(T4K_data@scale.data[T4K_data@var.genes, ]), membership = T4K_data@ident)
stats:::plot.dendrogram(T4K_hierarchy$dend)
```

Let's perform alignment on the above two trees. First, we uses `as_binary_tree` convert the `hclust` object to a `binary_tree` object that labels every node in the tree and compute the mean of each cluster. 

```{r message=FALSE, warning=FALSE}
T3K_tree <- with(T3K_hierarchy, ReSET::as_binary_tree(hclust, data, membership))
T4K_tree <- with(T4K_hierarchy, ReSET::as_binary_tree(hclust, data, membership))
```

Here is the same two trees with every node labeled. 

```{r message=FALSE, warning=FALSE, fig.width = 8, fig.height = 4}
par(mfrow = c(1, 2))
plot(T3K_tree$tree)
plot(T4K_tree$tree)
```

We create the cost matrix to perform alignment next. We use 1 - Pearson correlation coeffcient of every pair of nodes between the two trees. 

```{r message=FALSE, warning=FALSE, fig.width = 8, fig.height = 4}
cost_matrix <- ReSET::cost_matrix(T3K_tree$summary_stats, T4K_tree$summary_stats)
cost_matrix[1:5, 1:5]
```

To align the trees, we use the `align` function. We plot the aligned tree as follows. 

```{r message=FALSE, warning=FALSE, fig.width = 8, fig.height = 6}
align_obj <- ReSET::align(T3K_tree$tree, T4K_tree$tree, cost_matrix)
plot(align_obj$tree)
```

## Customized clustering

In the following example, we show how to use `ReSET` with user-provided clustering. On the same dataset, we performed clustering from other pipeline and manually labeled each cluster. The normalized data and cluster labels are provided in `T3K` and `T4K` dataset. First, we construct the hierarchy. 

```{r message=FALSE, warning=FALSE}
data('T3K')
T3K_tree <- ReSET::construct_hierarchy(t(T3K$normalized_data), T3K$membership)
data('T4K')
T4K_tree <- ReSET::construct_hierarchy(t(T4K$normalized_data), T4K$membership)
```

Here are the constructed hierarchies:

```{r message=FALSE, warning=FALSE, fig.width = 8, fig.height = 5}
par(mfrow = c(1, 2))
stats:::plot.dendrogram(T3K_tree$dend)
stats:::plot.dendrogram(T4K_tree$dend)
```


We then convert the object into `binary_tree` objects and summarize each node by its median expression.

```{r message=FALSE, warning=FALSE}
T3K_binary_tree <- with(T3K_tree, ReSET::as_binary_tree(hclust, data, membership, median))
T4K_binary_tree <- with(T4K_tree, ReSET::as_binary_tree(hclust, data, membership, median))
```


We defined the cost matrix similarly and run alignment algorithm. 

```{r message=FALSE, warning=FALSE}
cost_matrix <- ReSET::cost_matrix(T3K_binary_tree$summary_stats, T4K_binary_tree$summary_stats)
align_obj <- ReSET::align(T3K_binary_tree$tree, T4K_binary_tree$tree, cost_matrix)
```

The two trees and aligned trees are visualized below. 

```{r message=FALSE, warning=FALSE, fig.width = 8, fig.height = 5}
par(mfrow = c(1, 2))
plot(T3K_binary_tree$tree)
plot(T4K_binary_tree$tree)
```

```{r message=FALSE, warning=FALSE, fig.width = 8, fig.height = 5}
par(mfrow = c(1, 1))
plot(align_obj$tree)
```

