---
title: "Prepare the data"
author: "Meng Wang"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  \usepackage[utf8]{inputenc}
---

The purpose of this vignette demonstrates how to prepare data for ReSET. First, we read in and preprocess two 10X scRNAseq datasets containing around 3000 and 4000 Pan T cells from a single normal donor. The two datasets are available [here](https://support.10xgenomics.com/single-cell-gene-expression/datasets/2.1.0/t_3k) and [here](https://support.10xgenomics.com/single-cell-gene-expression/datasets/2.1.0/t_4k). We used `preprocess_data` function to filter and normalize the data. PCA were performed as well. The output is a Seurat object.

```{r message=FALSE, warning=FALSE, include=FALSE}
T3K_path <- system.file("extdata", '3k_pan_T_filtered_gene_bc_matrices/GRCh38', package = 'ReSET', mustWork = T)
T4K_path <- system.file("extdata", "4k_pan_T_filtered_gene_bc_matrices/GRCh38", package = 'ReSET', mustWork = T)

T3K_data <- ReSET::preprocess_data(T3K_path)
T4K_data <- ReSET::preprocess_data(T4K_path)
```

We performed graph-based clustering using `Seurat` and construct a tree of the clusters using hierarchical clustering on the mean of gene expression. Here is the dendrogram for `T3K` dataset.

```{r message=FALSE, warning=FALSE, fig.width = 6, fig.height = 3}
T3K_data <- Seurat::FindClusters(T3K_data, reduction.type = "pca", dims.use = 1:12, resolution = 2, print.output = 0) # 12 clusters
T3K_hierarchy <- ReSET::construct_hierarchy(T3K_data@scale.data, membership = T3K_data@ident, mean)
plot(T3K_hierarchy$dend)
```

And here is the dendrogram for `T4K` dataset.

```{r message=FALSE, warning=FALSE, fig.width = 6, fig.height = 3}
T4K_data <- Seurat::FindClusters(T4K_data, reduction.type = "pca", dims.use = 1:12, resolution = 2, print.output = 0) # 13 clusters
T4K_hierarchy <- ReSET::construct_hierarchy(T4K_data@scale.data, membership = T4K_data@ident, mean)
plot(T4K_hierarchy$dend)
```